File: sdks/nodejs/src/integrations/firecrawl.js
import axios from 'axios';
import { emitAPICallEvent } from '../utilities/sanitizeAPICall.js';

export default class FirecrawlIntegration {
    constructor(apiKey, options = {}) {
        this.apiKey = apiKey;
        this.options = {
            baseURL: 'https://api.firecrawl.dev/v2',
            timeout: 60000, // Longer timeout for scraping operations
            ...options
        };
    }

    /**
     * Scrape a single URL with various options
     * @param {Object} params - Scraping parameters
     * @returns {Promise<Object>} Scraping response
     */
    async scrape(params = {}) {
        try {
            const url = `${this.options.baseURL}/scrape`;

            // Remove any params that are null, undefined, or empty strings
            Object.keys(params).forEach(key => {
                if (params[key] === '') {
                    delete params[key];
                }
                if (params[key] === null) {
                    delete params[key];
                }
                if (params[key] === undefined) {
                    delete params[key];
                }
                if (Array.isArray(params[key]) && params[key].length === 0) {
                    delete params[key];
                }
            });

            const requestHeaders = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${this.apiKey}` };
            const startTime = Date.now();

            const response = await axios({
                url: url,
                method: 'POST',
                headers: requestHeaders,
                data: params,
                timeout: this.options.timeout
            });

            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'firecrawl', nodeId: null, nodeType: null,
                request: { method: 'POST', url, headers: requestHeaders, body: params },
                response: { status: response.status, statusText: response.statusText },
                duration: Date.now() - startTime, error: null
            });

            if (response.status >= 400) {
                throw new Error(`Firecrawl API error: ${response.status} - ${response.data?.error || response.statusText}`);
            }

            return response.data;

        } catch (error) {
            // Note: emitAPICallEvent for error case is handled by the catch in the caller
            // since errors from axios.post throw before we can capture the response
            if (error.response) {
                const status = error.response.status;
                const statusText = error.response.statusText;
                const responseData = error.response.data;
                
                let errorMessage = `Firecrawl API Error (${status} ${statusText})`;
                if (responseData?.error) {
                    errorMessage += `: ${responseData.error}`;
                }
                
                throw new Error(errorMessage);
            } else if (error.request) {
                throw new Error('Firecrawl API Error: No response received');
            } else {
                throw new Error(`Firecrawl API Error: ${error.message}`);
            }
        }
    }

    /**
     * Search the web with the Firecrawl Search API
     * @param {Object} params - Search parameters (query, limit, ...)
     * @returns {Promise<Object>} Search response
     */
    async search(params = {}) {
        try {
            const url = `${this.options.baseURL}/search`;

            // Remove any params that are null, undefined, or empty strings
            Object.keys(params).forEach(key => {
                if (params[key] === '') {
                    delete params[key];
                }
                if (params[key] === null) {
                    delete params[key];
                }
                if (params[key] === undefined) {
                    delete params[key];
                }
                if (Array.isArray(params[key]) && params[key].length === 0) {
                    delete params[key];
                }
            });

            const requestHeaders = { 'Content-Type': 'application/json', 'Authorization': `Bearer ${this.apiKey}` };
            const startTime = Date.now();

            const response = await axios({
                url: url,
                method: 'POST',
                headers: requestHeaders,
                data: params,
                timeout: this.options.timeout
            });

            await emitAPICallEvent(this._engineConfig, {
                timestamp: startTime, integration: 'firecrawl', nodeId: null, nodeType: null,
                request: { method: 'POST', url, headers: requestHeaders, body: params },
                response: { status: response.status, statusText: response.statusText },
                duration: Date.now() - startTime, error: null
            });

            if (response.status >= 400) {
                throw new Error(`Firecrawl API error: ${response.status} - ${response.data?.error || response.statusText}`);
            }

            return response.data;

        } catch (error) {
            // Note: emitAPICallEvent for error case is handled by the catch in the caller
            // since errors from axios.post throw before we can capture the response
            if (error.response) {
                const status = error.response.status;
                const statusText = error.response.statusText;
                const responseData = error.response.data;

                let errorMessage = `Firecrawl API Error (${status} ${statusText})`;
                if (responseData?.error) {
                    errorMessage += `: ${responseData.error}`;
                }

                throw new Error(errorMessage);
            } else if (error.request) {
                throw new Error('Firecrawl API Error: No response received');
            } else {
                throw new Error(`Firecrawl API Error: ${error.message}`);
            }
        }
    }

    /**
     * Helper function to convert comma-separated string to array
     * @param {string|Array} input - String or array input
     * @returns {Array} Array of strings
     */
    static stringToArray(input) {
        if (Array.isArray(input)) {
            return input;
        }
        if (typeof input === 'string') {
            return input.split(',').map(item => item.trim()).filter(item => item.length > 0);
        }
        return [];
    }

    /**
     * Helper function to build query parameters with array handling
     * @param {Object} params - Raw parameters
     * @returns {Object} Processed parameters
     */
    static processParams(params) {
        const processed = { ...params };
        
        // Handle parameter name mappings (snake_case input names to API parameter names)
        const paramMappings = {
            'include_tags': 'includeTags',
            'exclude_tags': 'excludeTags',
            'only_main_content': 'onlyMainContent',
            'max_age': 'maxAge',
            'wait_for': 'waitFor',
            'mobile_device': 'mobile',
            'skip_tls_verification': 'skipTlsVerification',
            'remove_base64_images': 'removeBase64Images',
            'block_ads': 'blockAds',
            'store_in_cache': 'storeInCache',
            'zero_data_retention': 'zeroDataRetention'
        };
        
        // Apply parameter name mappings
        Object.entries(paramMappings).forEach(([inputKey, apiKey]) => {
            if (processed[inputKey] !== undefined) {
                processed[apiKey] = processed[inputKey];
                delete processed[inputKey]; // Remove the old key
            }
        });
        
        // Fields that should be converted from comma-separated strings to arrays
        const arrayFields = [
            'includeTags', 'excludeTags', 'formats'
        ];

        arrayFields.forEach(field => {
            if (processed[field] !== undefined) {
                const arrayValue = this.stringToArray(processed[field]);
                if (arrayValue.length > 0) {
                    processed[field] = arrayValue;
                }
            }
        });

        // Handle formats - convert string to array of objects if needed
        if (processed.formats && Array.isArray(processed.formats)) {
            processed.formats = processed.formats.map(format => {
                if (typeof format === 'string') {
                    return { type: format };
                }
                return format;
            });
        }

        return processed;
    }
}

File: sdks/python/src/integrations/firecrawl.py
"""
Firecrawl Integration for the zv1 engine.

Provides web scraping capabilities through the Firecrawl API.
"""

from __future__ import annotations

import logging
import time
from typing import Any

from src.utilities.sanitize_api_call import emit_api_call_event

try:
    import httpx

    HAS_HTTPX = True
except ImportError:
    HAS_HTTPX = False

logger = logging.getLogger(__name__)


class FirecrawlIntegration:
    """
    Integration with Firecrawl's web scraping API.
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.firecrawl.dev/v2",
        timeout: float = 60.0,
    ) -> None:
        """
        Initialize the Firecrawl integration.

        Args:
            api_key: Firecrawl API key.
            base_url: Base URL for the API.
            timeout: Request timeout in seconds (default 60s for scraping).
        """
        if not HAS_HTTPX:
            raise ImportError(
                "httpx package is required for Firecrawl integration. "
                "Install with: pip install httpx"
            )

        self.api_key = api_key
        self.base_url = base_url
        self.timeout = timeout

        self.client = httpx.AsyncClient(
            base_url=base_url,
            timeout=timeout,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {api_key}",
            },
        )

    async def scrape(self, params: dict[str, Any]) -> dict[str, Any]:
        """
        Scrape a single URL with various options.

        Args:
            params: Scraping parameters (url, formats, etc.).

        Returns:
            Scraping response.

        Raises:
            Exception: On API errors.
        """
        # Remove empty params
        clean_params = self._clean_params(params)

        start_time = int(time.time() * 1000)
        request_url = f"{self.base_url}/scrape"

        try:
            response = await self.client.post("/scrape", json=clean_params)

            await emit_api_call_event(getattr(self, "_engine_config", None), {
                "timestamp": start_time, "integration": "firecrawl", "nodeId": None, "nodeType": None,
                "request": {"method": "POST", "url": request_url, "headers": dict(self.client.headers), "body": clean_params},
                "response": {"status": response.status_code, "statusText": response.reason_phrase or ""},
                "duration": int(time.time() * 1000) - start_time, "error": None,
            })

            if response.status_code >= 400:
                error_data = response.json() if response.content else {}
                error_msg = error_data.get("error", response.reason_phrase)
                raise Exception(f"Firecrawl API error: {response.status_code} - {error_msg}")

            return response.json()

        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            status_text = e.response.reason_phrase
            try:
                response_data = e.response.json()
                error_detail = response_data.get("error", "")
            except Exception:
                error_detail = ""

            error_message = f"Firecrawl API Error ({status} {status_text})"
            if error_detail:
                error_message += f": {error_detail}"

            raise Exception(error_message) from e

        except httpx.RequestError as e:
            raise Exception(f"Firecrawl API Error: {e}") from e

    async def search(self, params: dict[str, Any]) -> dict[str, Any]:
        """
        Search the web with the Firecrawl Search API.

        Args:
            params: Search parameters (query, limit, ...).

        Returns:
            Search response.

        Raises:
            Exception: On API errors.
        """
        # Remove empty params
        clean_params = self._clean_params(params)

        start_time = int(time.time() * 1000)
        request_url = f"{self.base_url}/search"

        try:
            response = await self.client.post("/search", json=clean_params)

            await emit_api_call_event(getattr(self, "_engine_config", None), {
                "timestamp": start_time, "integration": "firecrawl", "nodeId": None, "nodeType": None,
                "request": {"method": "POST", "url": request_url, "headers": dict(self.client.headers), "body": clean_params},
                "response": {"status": response.status_code, "statusText": response.reason_phrase or ""},
                "duration": int(time.time() * 1000) - start_time, "error": None,
            })

            if response.status_code >= 400:
                error_data = response.json() if response.content else {}
                error_msg = error_data.get("error", response.reason_phrase)
                raise Exception(f"Firecrawl API error: {response.status_code} - {error_msg}")

            return response.json()

        except httpx.HTTPStatusError as e:
            status = e.response.status_code
            status_text = e.response.reason_phrase
            try:
                response_data = e.response.json()
                error_detail = response_data.get("error", "")
            except Exception:
                error_detail = ""

            error_message = f"Firecrawl API Error ({status} {status_text})"
            if error_detail:
                error_message += f": {error_detail}"

            raise Exception(error_message) from e

        except httpx.RequestError as e:
            raise Exception(f"Firecrawl API Error: {e}") from e

    def _clean_params(self, params: dict[str, Any]) -> dict[str, Any]:
        """Remove empty/null params."""
        clean = {}
        for key, value in params.items():
            if value is None:
                continue
            if value == "":
                continue
            if isinstance(value, list) and len(value) == 0:
                continue
            clean[key] = value
        return clean

    @staticmethod
    def string_to_array(input_val: str | list[str]) -> list[str]:
        """
        Convert comma-separated string to array.

        Args:
            input_val: String or array input.

        Returns:
            Array of strings.
        """
        if isinstance(input_val, list):
            return input_val
        if isinstance(input_val, str):
            return [item.strip() for item in input_val.split(",") if item.strip()]
        return []

    @staticmethod
    def process_params(params: dict[str, Any]) -> dict[str, Any]:
        """
        Process parameters with array handling and name mappings.

        Args:
            params: Raw parameters.

        Returns:
            Processed parameters.
        """
        processed = dict(params)

        # Parameter name mappings (snake_case to camelCase)
        param_mappings = {
            "include_tags": "includeTags",
            "exclude_tags": "excludeTags",
            "only_main_content": "onlyMainContent",
            "max_age": "maxAge",
            "wait_for": "waitFor",
            "mobile_device": "mobile",
            "skip_tls_verification": "skipTlsVerification",
            "remove_base64_images": "removeBase64Images",
            "block_ads": "blockAds",
            "store_in_cache": "storeInCache",
            "zero_data_retention": "zeroDataRetention",
        }

        # Apply parameter name mappings
        for input_key, api_key in param_mappings.items():
            if input_key in processed:
                processed[api_key] = processed.pop(input_key)

        # Fields that should be converted from comma-separated strings to arrays
        array_fields = ["includeTags", "excludeTags", "formats"]

        for field in array_fields:
            if field in processed:
                array_value = FirecrawlIntegration.string_to_array(processed[field])
                if array_value:
                    processed[field] = array_value

        # Handle formats - convert string to array of objects if needed
        if "formats" in processed and isinstance(processed["formats"], list):
            processed["formats"] = [
                {"type": fmt} if isinstance(fmt, str) else fmt
                for fmt in processed["formats"]
            ]

        return processed

    async def close(self) -> None:
        """Close the HTTP client."""
        await self.client.aclose()
